demo/actions.py

```diff
@@ -631,3 +635,141 @@ def run(self, dispatcher, tracker, domain):
+                )
+            )
+        )
+
+        return has_instance and not excluded
+
+    filtered = filter(filter_function, reversed(tracker.events))
+    for i in range(skip):
+        next(filtered, None)
+
+    logger.error("filtered returning: {}".format(filtered))
+    return next(filtered, None)
+
+
+def GetAlgoliaLink(hits, index: int):
+    doc_link = "- [" + hits[index]["hierarchy"]["lvl0"]
+    if hits[index]["hierarchy"]["lvl1"]:
+        doc_link += "/" + hits[index]["hierarchy"]["lvl1"].strip()
+        if hits[index]["hierarchy"]["lvl2"]:
+            doc_link += "/" + hits[index]["hierarchy"]["lvl2"].strip()
+    doc_link += "](" + hits[index]["url"] + ")"
+    return doc_link
+
+
+class ActionDocsSearch(Action):
+    def name(self):
+        return "action_docs_search"
+
+    def __init__(self) -> None:
+        self.algolia = AlgoliaAPI(
+            config.algolia_app_id, config.algolia_search_key, config.algolia_docs_index
+        )
+
+    def run(self, dispatcher, tracker, domain):
+        search_text = tracker.latest_message["text"]
+        logger.error("Doing doc search on test: {}".format(search_text))
+        if search_text == "/technical_question{}":
+            last_user_event = get_last_event_for(tracker, "user", skip=2)
+            logger.error("last_user_event: {}".format(last_user_event))
+            search_text = last_user_event["text"]
+        logger.error("search_text: {}".format(search_text))
+
+        # Search of docs pages
+        alg_res = self.algolia.search(search_text)
+        logger.error(
+            "alg_res[hits][0][hierarchy]: {}".format(alg_res["hits"][0]["hierarchy"])
+        )
+
+        doc_list = GetAlgoliaLink(alg_res["hits"], 0)
+        doc_list += "\n" + GetAlgoliaLink(alg_res["hits"], 1)
+
+        logger.error("ActionDocsSearch, docs: {}".format(doc_list))
+        dispatcher.utter_message(
+            "I can't answer your question directly, but I found the following from the docs:\n"
+            + doc_list
+        )
+
+        return []
+
+
+def GetDiscourseLinks(topics, index: int):
+    doc_url = (
+        "https://forum.rasa.com/t/"
+        + topics[index]["slug"]
+        + "/"
+        + str(topics[index]["id"])
+    )
+    logger.error("ActionDocsForumSearch, doc_url: {}".format(doc_url))
+    forum = "- [" + topics[index]["title"] + "](" + doc_url + ")"
+    return forum
+
+
+class ActionForumSearch(Action):
+    def name(self):
+        return "action_forum_search"
+
+    def __init__(self) -> None:
+        self.discourse = DiscourseAPI("https://forum.rasa.com/search")
+
+    def run(self, dispatcher, tracker, domain):
+        search_text = tracker.latest_message["text"]
+        if search_text == "/technical_question{}":
+            last_user_event = get_last_event_for(tracker, "user", skip=2)
+            logger.error("last_user_event: {}".format(last_user_event))
+            search_text = last_user_event["text"]
```

Comment:
whats happening here?

Reply:
We could get here via `action_default_ask_affirmation` in which case we have to look back to get the user utterance from the previous intent.